File: joonas/081.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    vector<stack<int>> s(board.size());
    for (int r = board.size()-1; r >= 0; --r) {
        for (int c = 0; c < board[r].size(); ++c) {
            if (board[r][c] > 0) s[c].push(board[r][c]);
        }
    }
    stack<int> z;
    int answer = 0;
    for (auto& m : moves) {
        if (s[m-1].empty()) continue;
        if (!z.empty() && s[m-1].top() == z.top()) {
            z.pop();
            answer += 2;
        } else {
            z.push(s[m-1].top());
        }
        s[m-1].pop();
    }
    return answer;
}

File: jlogkim/081.java
import java.util.Stack;

class Solution {
    public int solution(int[][] board, int[] moves) {
       int answer = 0;
        int n = board.length;
        int ind[] = new int[n];
        int i;
        for(int c=0;c<n;c++) {
	        for(i=0;i<n;i++) {
	        	if(board[i][c]>0) break;
	        } ind[c]=i;
	        System.out.print(ind[c]+" ");
        }
        System.out.println();
        Stack<Integer> stack=new Stack<Integer>();
        
        int totalmove = moves.length;
        int targetc;
        for( i=0; i<totalmove; i++) {
        	targetc=moves[i]-1;
        	if(ind[targetc]<n) {
        		if(!stack.isEmpty()&&stack.peek()==board[ind[targetc]][targetc]) {
        			answer+=2;
        			stack.pop();
        		}else {
        			stack.push(board[ind[targetc]][targetc]);
        		}
        		ind[moves[i]-1]++;
        	}
        }
        
        return answer;
    }
}